### PR TITLE
Making matrix tests more reliable

### PR DESCRIPTION
**Pull request by IanCal**
Check we're in the right rooms when operating mostly.

I have added a change to where the room name is rendered, but I don't think we should have it there. It dealt with an issue where the *name* of the room in the rooms-manager had updated but the room *itself* had not yet rendered.

Broadly there's a few places we are not checking we're actually in the right room yet before sending messages. Sometimes we're checking if the room is settled, but it's executing *before* anything happens rather than checking they're finished. We may be able to use the settled count to identify this (you can check the number of times a task has run).

To cause more errors, run playwright in parallel, I get more flaky tests as the workers go up to 32.

### CODE REVIEW

File: packages/host/app/components/matrix/room.gts [hunk 1]
```diff
@@ -41,6 +41,9 @@ interface RoomArgs {
 
 export default class Room extends Component<RoomArgs> {
   <template>
+    <!-- this is part of testing that the actual component is now rendered, -->
+    <div data-test-rooms-header={{this.room.name}}>Room:
```
Comment (IanCal) [inline packages/host/app/components/matrix/room.gts]: This shouldn't really be dealt with here.

File: packages/host/app/components/matrix/room.gts [hunk 2]
```diff
@@ -126,7 +129,7 @@ export default class Room extends Component<RoomArgs> {
 
     <div class='send-message'>
       <BoxelInput
-        data-test-message-field
+        data-test-message-field={{this.room.name}}
```
Comment (IanCal) [inline packages/host/app/components/matrix/room.gts:130]: Make sure that we're looking at the correct send message field.

File: packages/matrix/helpers/index.ts [hunk 3]
```diff
@@ -119,6 +119,12 @@ export async function createRoom(
       .fill(roomDetails.invites.join(', '));
   }
   await page.locator('[data-test-create-room-btn]').click();
+  await isInRoom(page, roomDetails.name);
+}
+
+export async function isInRoom(page: Page, roomName: string) {
+  await page.locator(`[data-test-rooms-header="${roomName}"]`).waitFor();
+  await expect(page.locator(`[data-test-room-settled]`)).toHaveCount(1);
```
Comment (IanCal) [inline packages/matrix/helpers/index.ts:127]: Check we're in a settled state after we've checked we're in the correct room.

File: packages/matrix/helpers/index.ts [hunk 4]
```diff
@@ -131,7 +137,18 @@ export async function leaveRoom(page: Page, roomName: string) {
...
-  await expect(page.locator(`[data-test-room-settled]`)).toHaveCount(1);
+  await isInRoom(page, roomName);
+}
+
+export async function writeMessage(
+  page: Page,
+  roomName: string,
+  message: string,
+) {
+  await page.locator(`[data-test-message-field="${roomName}"]`).fill(message);
```
Comment (IanCal) [inline packages/matrix/helpers/index.ts:148]: Some tests were writing messages in the field for the previous room as it happened too fast.

The header seemed to change before this element was changed.

**Comment by IanCal** [comment]
Sorted the tests, last thing is to try removing the extra room name that got added in.

**Comment by habdelra** [comment]
I like your updates

**Review by habdelra — APPROVED** [review]

**Comment by habdelra** [comment]
I'm unsure why running in parallel would reveal more flakiness. presumably we are just doing reads from the realm server which it should have no issues with for parallel access. unsure what the nature of the parallel issues are...

**Comment by IanCal** [comment]
> I'm unsure why running in parallel would reveal more flakiness. presumably we are just doing reads from the realm server which it should have no issues with for parallel access. unsure what the nature of the parallel issues are...

It leans more towards timeouts I think. Sometimes just *nothing* loads. I'm not sure if this is a speed issue, or it could also be down to networking things. When there's lots of starting and stopping of docker containers I sometimes see connection issues while browsing, so it might be causing the flakiness and be nothing to do with our work. It tends to happen *roughly* when I have as many parallel tests as real cores, which feels a bit suspicious but not exact enough to track down.

**Pull request merged**